File: bindings/assemblyscript/assembly/fs.ts
import { Result } from "./result";
import {
  ERROR_RESULT_VALUE_OFFSET_4,
  ERROR_RESULT_VALUE_OFFSET_8,
  Oid,
  RESULT_ERROR_SIZE,
  alloc,
  liftError,
  liftString,
  resultIsOk,
  utf8Bytes,
} from "./wit";

export const FS_O_RDONLY: u32 = 0;
export const FS_O_WRONLY: u32 = 1;
export const FS_O_RDWR: u32 = 2;

export const FS_SEEK_SET: u32 = 0;
export const FS_SEEK_CUR: u32 = 1;
export const FS_SEEK_END: u32 = 2;

const FS_STAT_ENTRY_PTR_OFFSET: usize = 24;
const FS_STAT_ENTRY_LEN_OFFSET: usize = 28;
const FS_STAT_LENGTH_OFFSET: usize = 32;
const FS_STAT_STATE_OFFSET: usize = 40;
const FS_STAT_RESULT_SIZE: usize = 56;
const FS_DIRENT_NAME_LEN_OFFSET: usize = 4;
const FS_DIRENT_IS_DIR_OFFSET: usize = 8;
const FS_DIRENT_LENGTH_OFFSET: usize = 16;
const FS_DIRENT_SIZE: usize = 24;

export class FsStat {
  oid: Oid;
  generation: u64;
  entry: string;
  length: u64;
  state: u32;

  constructor(oid: Oid = new Oid(), generation: u64 = 0, entry: string = "", length: u64 = 0, state: u32 = 0) {
    this.oid = oid;
    this.generation = generation;
    this.entry = entry;
    this.length = length;
    this.state = state;
  }
}

export class FsDirEntry {
  name: string;
  isDir: bool;
  length: u64;

  constructor(name: string = "", isDir: bool = false, length: u64 = 0) {
    this.name = name;
    this.isDir = isDir;
    this.length = length;
  }
}

@external("mududb:component-shim/system", "fs-open")
declare function rawFsOpen(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, pathPtr: usize, pathLen: usize, flags: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-close")
declare function rawFsClose(sessionHi: u64, sessionLo: u64, fd: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-read")
declare function rawFsRead(sessionHi: u64, sessionLo: u64, fd: u32, len: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-write")
declare function rawFsWrite(sessionHi: u64, sessionLo: u64, fd: u32, dataPtr: usize, dataLen: usize, result: usize): void;
@external("mududb:component-shim/system", "fs-pread")
declare function rawFsPread(sessionHi: u64, sessionLo: u64, fd: u32, offset: u64, len: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-pwrite")
declare function rawFsPwrite(sessionHi: u64, sessionLo: u64, fd: u32, offset: u64, dataPtr: usize, dataLen: usize, result: usize): void;
@external("mududb:component-shim/system", "fs-lseek")
declare function rawFsLseek(sessionHi: u64, sessionLo: u64, fd: u32, offset: i64, whence: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-fstat")
declare function rawFsFstat(sessionHi: u64, sessionLo: u64, fd: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-stat")
declare function rawFsStat(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, pathPtr: usize, pathLen: usize, result: usize): void;
@external("mududb:component-shim/system", "fs-fsync")
declare function rawFsFsync(sessionHi: u64, sessionLo: u64, fd: u32, result: usize): void;
@external("mududb:component-shim/system", "fs-readdir")
declare function rawFsReaddir(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, pathPtr: usize, pathLen: usize, result: usize): void;

function liftBuffer(ptr: usize, len: usize): ArrayBuffer {
  const out = new ArrayBuffer(<i32>len);
  memory.copy(changetype<usize>(out), ptr, len);
  return out;
}

function liftFsStat(ptr: usize): FsStat {
  return new FsStat(
    new Oid(load<u64>(ptr), load<u64>(ptr + 8)),
    load<u64>(ptr + 16),
    liftString(load<u32>(ptr + FS_STAT_ENTRY_PTR_OFFSET), load<u32>(ptr + FS_STAT_ENTRY_LEN_OFFSET)),
    load<u64>(ptr + FS_STAT_LENGTH_OFFSET),
    load<u32>(ptr + FS_STAT_STATE_OFFSET),
  );
}

function liftFsDirEntries(ptr: usize, len: u32): FsDirEntry[] {
  const entries: FsDirEntry[] = [];
  for (let i: u32 = 0; i < len; i++) {
    const base = ptr + <usize>i * FS_DIRENT_SIZE;
    entries.push(new FsDirEntry(
      liftString(load<u32>(base), load<u32>(base + FS_DIRENT_NAME_LEN_OFFSET)),
      load<u8>(base + FS_DIRENT_IS_DIR_OFFSET) != 0,
      load<u64>(base + FS_DIRENT_LENGTH_OFFSET),
    ));
  }
  return entries;
}

export function fsOpen(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, path: string, flags: u32): Result<u32> {
  const bytes = utf8Bytes(path);
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsOpen(sessionHi, sessionLo, oidHi, oidLo, changetype<usize>(bytes), bytes.byteLength, flags, out);
  return resultIsOk(out)
    ? Result.ok<u32>(load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4))
    : Result.error<u32>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsClose(sessionHi: u64, sessionLo: u64, fd: u32): Result<bool> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsClose(sessionHi, sessionLo, fd, out);
  return resultIsOk(out)
    ? Result.ok<bool>(true)
    : Result.error<bool>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsRead(sessionHi: u64, sessionLo: u64, fd: u32, len: u32): Result<ArrayBuffer> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsRead(sessionHi, sessionLo, fd, len, out);
  return resultIsOk(out)
    ? Result.ok<ArrayBuffer>(liftBuffer(load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4), load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4 + 4)))
    : Result.error<ArrayBuffer>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsWrite(sessionHi: u64, sessionLo: u64, fd: u32, data: ArrayBuffer): Result<u32> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsWrite(sessionHi, sessionLo, fd, changetype<usize>(data), data.byteLength, out);
  return resultIsOk(out)
    ? Result.ok<u32>(load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4))
    : Result.error<u32>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsPread(sessionHi: u64, sessionLo: u64, fd: u32, offset: u64, len: u32): Result<ArrayBuffer> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsPread(sessionHi, sessionLo, fd, offset, len, out);
  return resultIsOk(out)
    ? Result.ok<ArrayBuffer>(liftBuffer(load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4), load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4 + 4)))
    : Result.error<ArrayBuffer>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsPwrite(sessionHi: u64, sessionLo: u64, fd: u32, offset: u64, data: ArrayBuffer): Result<bool> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsPwrite(sessionHi, sessionLo, fd, offset, changetype<usize>(data), data.byteLength, out);
  return resultIsOk(out)
    ? Result.ok<bool>(true)
    : Result.error<bool>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsLseek(sessionHi: u64, sessionLo: u64, fd: u32, offset: i64, whence: u32): Result<u64> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsLseek(sessionHi, sessionLo, fd, offset, whence, out);
  return resultIsOk(out)
    ? Result.ok<u64>(load<u64>(out + ERROR_RESULT_VALUE_OFFSET_8))
    : Result.error<u64>(liftError(out + ERROR_RESULT_VALUE_OFFSET_8));
}

export function fsFstat(sessionHi: u64, sessionLo: u64, fd: u32): Result<FsStat> {
  const out = alloc(FS_STAT_RESULT_SIZE);
  rawFsFstat(sessionHi, sessionLo, fd, out);
  return resultIsOk(out)
    ? Result.ok<FsStat>(liftFsStat(out + ERROR_RESULT_VALUE_OFFSET_8))
    : Result.error<FsStat>(liftError(out + ERROR_RESULT_VALUE_OFFSET_8));
}

export function fsStat(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, path: string): Result<FsStat> {
  const bytes = utf8Bytes(path);
  const out = alloc(FS_STAT_RESULT_SIZE);
  rawFsStat(sessionHi, sessionLo, oidHi, oidLo, changetype<usize>(bytes), bytes.byteLength, out);
  return resultIsOk(out)
    ? Result.ok<FsStat>(liftFsStat(out + ERROR_RESULT_VALUE_OFFSET_8))
    : Result.error<FsStat>(liftError(out + ERROR_RESULT_VALUE_OFFSET_8));
}

export function fsFsync(sessionHi: u64, sessionLo: u64, fd: u32): Result<bool> {
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsFsync(sessionHi, sessionLo, fd, out);
  return resultIsOk(out)
    ? Result.ok<bool>(true)
    : Result.error<bool>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

export function fsReaddir(sessionHi: u64, sessionLo: u64, oidHi: u64, oidLo: u64, path: string): Result<FsDirEntry[]> {
  const bytes = utf8Bytes(path);
  const out = alloc(RESULT_ERROR_SIZE);
  rawFsReaddir(sessionHi, sessionLo, oidHi, oidLo, changetype<usize>(bytes), bytes.byteLength, out);
  return resultIsOk(out)
    ? Result.ok<FsDirEntry[]>(liftFsDirEntries(load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4), load<u32>(out + ERROR_RESULT_VALUE_OFFSET_4 + 4)))
    : Result.error<FsDirEntry[]>(liftError(out + ERROR_RESULT_VALUE_OFFSET_4));
}

File: crates/sdk/mudu_api/rust/src/mock/fs.rs
//! In-memory debug emulation of the host fs syscall family (message kinds
//! `FsOpen`..`FsReaddir`), mirroring the C# `MockFsEmulation`. Files live in
//! a flat `(oid, entry) -> bytes` map and file descriptors in a min-free-u32
//! table with a per-fd cursor. The emulation deliberately implements a small
//! POSIX subset:
//!
//! - open access modes 0 (read-only), 1 (write-only) and 2 (read/write); any
//!   other flag bit (`O_CREAT`/`O_TRUNC`/`O_APPEND`/`O_EXCL` style bits) or
//!   the reserved mode value 3 is rejected with errno 22 (`EINVAL`);
//! - a read open of a missing entry fails with errno 2 (`ENOENT`); a write
//!   open creates the entry and truncates it (debug emulation of
//!   `O_CREAT|O_TRUNC`);
//! - read/pread clamp at EOF; write/pwrite extend the file and zero-fill any
//!   sparse gap; pread/pwrite do not move the fd cursor;
//! - lseek supports whence 0/1/2 (`SET`/`CUR`/`END`); an unknown whence, a
//!   negative result or an overflow is errno 22;
//! - fstat/stat report `{ generation = 1, length, state = 1 }`; stat of a
//!   missing entry is errno 2;
//! - readdir lists the immediate children of the path within one fs object;
//!   readdir on a file path is errno 20 (`ENOTDIR`);
//! - operations on an unknown fd fail with errno 9 (`EBADF`); fsync is a
//!   no-op.
//!
//! All state is static, process-wide debug state: nothing is persisted. The
//! fd table is per-process, mirroring the host's min-free fd allocation
//! starting at 0.

use crate::types::UniReturn;
use crate::universal::uni_error::UniError;
use crate::universal::uni_fs_dirent::UniFsDirent;
use crate::universal::uni_fs_open_argv::UniFsOpenArgv;
use crate::universal::uni_fs_stat::UniFsStat;
use crate::universal::uni_oid::UniOid;
use std::collections::{BTreeMap, HashMap};
use std::sync::{OnceLock, RwLock};

const ERRNO_NO_ENT: u32 = 2; // ENOENT
const ERRNO_BAD_F: u32 = 9; // EBADF
const ERRNO_NOT_DIR: u32 = 20; // ENOTDIR
const ERRNO_INVAL: u32 = 22; // EINVAL

const ACCESS_MODE_MASK: u32 = 3;
const MODE_READ_ONLY: u32 = 0;
const MODE_READ_WRITE: u32 = 2;

const EMULATED_GENERATION: u64 = 1;
const EMULATED_STATE: u32 = 1; // SEALED

type ObjectKey = (u64, u64);
type FileKey = (ObjectKey, String);

struct OpenFile {
    oid: ObjectKey,
    entry: String,
    cursor: u64,
}

fn files() -> &'static RwLock<HashMap<FileKey, Vec<u8>>> {
    static FILES: OnceLock<RwLock<HashMap<FileKey, Vec<u8>>>> = OnceLock::new();
    FILES.get_or_init(|| RwLock::new(HashMap::new()))
}

fn fd_table() -> &'static RwLock<HashMap<u32, OpenFile>> {
    static FD_TABLE: OnceLock<RwLock<HashMap<u32, OpenFile>>> = OnceLock::new();
    FD_TABLE.get_or_init(|| RwLock::new(HashMap::new()))
}

fn oid_key(oid: &UniOid) -> ObjectKey {
    (oid.h, oid.l)
}

fn error<T>(errno: u32, message: String) -> UniReturn<T> {
    Err(UniError {
        err_code: errno,
        err_msg: message,
        err_src: "MockFsEmulation".to_string(),
        ..Default::default()
    })
}

/// Allocates the smallest free fd, mirroring the host's allocation.
fn alloc_fd(table: &HashMap<u32, OpenFile>) -> u32 {
    let mut fd = 0;
    while table.contains_key(&fd) {
        fd += 1;
    }
    fd
}

/// Writes `data` at `offset`, growing the content and zero-filling any
/// sparse gap.
fn write_at(content: &mut Vec<u8>, offset: u64, data: &[u8]) {
    let offset = offset as usize;
    let end = offset + data.len();
    if end > content.len() {
        content.resize(end, 0);
    }
    content[offset..end].copy_from_slice(data);
}

fn stat(oid: ObjectKey, entry: &str) -> UniReturn<UniFsStat> {
    let files = files().read().expect("mock fs files lock poisoned");
    let Some(content) = files.get(&(oid, entry.to_string())) else {
        return error(ERRNO_NO_ENT, format!("fs-stat: no such entry '{entry}'"));
    };
    Ok(UniFsStat {
        oid: UniOid { h: oid.0, l: oid.1 },
        generation: EMULATED_GENERATION,
        entry: entry.to_string(),
        length: content.len() as u64,
        state: EMULATED_STATE,
    })
}

pub(super) fn fs_open(argv: UniFsOpenArgv) -> UniReturn<u32> {
    let mode = argv.flags & ACCESS_MODE_MASK;
    if (argv.flags & !ACCESS_MODE_MASK) != 0 || mode > MODE_READ_WRITE {
        return error(
            ERRNO_INVAL,
            format!("fs-open: unsupported flags 0x{:X}", argv.flags),
        );
    }

    let key = (oid_key(&argv.oid), argv.path.clone());
    let mut files = files().write().expect("mock fs files lock poisoned");
    if mode == MODE_READ_ONLY {
        if !files.contains_key(&key) {
            return error(
                ERRNO_NO_ENT,
                format!("fs-open: no such entry '{}'", argv.path),
            );
        }
    } else {
        // Debug emulation: a write open creates the entry and truncates it.
        files.insert(key, Vec::new());
    }
    drop(files);

    let mut table = fd_table().write().expect("mock fs fd table lock poisoned");
    let fd = alloc_fd(&table);
    table.insert(
        fd,
        OpenFile {
            oid: oid_key(&argv.oid),
            entry: argv.path,
            cursor: 0,
        },
    );
    Ok(fd)
}

pub(super) fn fs_close(fd: u32) -> UniReturn<()> {
    let mut table = fd_table().write().expect("mock fs fd table lock poisoned");
    if table.remove(&fd).is_none() {
        return error(ERRNO_BAD_F, format!("fs-close: unknown fd {fd}"));
    }
    Ok(())
}

/// Reads up to `len` bytes clamped at EOF from `data` starting at `start`.
fn read_clamped(data: &[u8], start: u64, len: u32) -> Vec<u8> {
    let start = start.min(data.len() as u64) as usize;
    let count = (len as usize).min(data.len() - start);
    data[start..start + count].to_vec()
}

pub(super) fn fs_read(fd: u32, len: u32) -> UniReturn<Vec<u8>> {
    let mut table = fd_table().write().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get_mut(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-read: unknown fd {fd}"));
    };
    let files = files().read().expect("mock fs files lock poisoned");
    let data = files
        .get(&(open.oid, open.entry.clone()))
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    let out = read_clamped(data, open.cursor, len);
    open.cursor += out.len() as u64;
    Ok(out)
}

pub(super) fn fs_write(fd: u32, data: Vec<u8>) -> UniReturn<u32> {
    let mut table = fd_table().write().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get_mut(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-write: unknown fd {fd}"));
    };
    let mut files = files().write().expect("mock fs files lock poisoned");
    let content = files.entry((open.oid, open.entry.clone())).or_default();
    write_at(content, open.cursor, &data);
    open.cursor += data.len() as u64;
    Ok(data.len() as u32)
}

pub(super) fn fs_pread(fd: u32, offset: u64, len: u32) -> UniReturn<Vec<u8>> {
    let table = fd_table().read().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-pread: unknown fd {fd}"));
    };
    let files = files().read().expect("mock fs files lock poisoned");
    let data = files
        .get(&(open.oid, open.entry.clone()))
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    Ok(read_clamped(data, offset, len))
}

pub(super) fn fs_pwrite(fd: u32, offset: u64, data: Vec<u8>) -> UniReturn<()> {
    let table = fd_table().read().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-pwrite: unknown fd {fd}"));
    };
    let mut files = files().write().expect("mock fs files lock poisoned");
    let content = files.entry((open.oid, open.entry.clone())).or_default();
    write_at(content, offset, &data);
    Ok(())
}

pub(super) fn fs_lseek(fd: u32, offset: i64, whence: u32) -> UniReturn<u64> {
    let mut table = fd_table().write().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get_mut(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-lseek: unknown fd {fd}"));
    };
    let files = files().read().expect("mock fs files lock poisoned");
    let length = files
        .get(&(open.oid, open.entry.clone()))
        .map_or(0, Vec::len) as i64;
    let base = match whence {
        0 => 0,                  // SEEK_SET
        1 => open.cursor as i64, // SEEK_CUR
        2 => length,             // SEEK_END
        other => {
            return error(ERRNO_INVAL, format!("fs-lseek: unknown whence {other}"));
        }
    };
    let Some(position) = base.checked_add(offset) else {
        return error(ERRNO_INVAL, "fs-lseek: position overflow".to_string());
    };
    if position < 0 {
        return error(
            ERRNO_INVAL,
            format!("fs-lseek: negative position {position}"),
        );
    }
    open.cursor = position as u64;
    Ok(position as u64)
}

pub(super) fn fs_fstat(fd: u32) -> UniReturn<UniFsStat> {
    let table = fd_table().read().expect("mock fs fd table lock poisoned");
    let Some(open) = table.get(&fd) else {
        return error(ERRNO_BAD_F, format!("fs-fstat: unknown fd {fd}"));
    };
    let entry = open.entry.clone();
    let oid = open.oid;
    drop(table);
    stat(oid, &entry)
}

pub(super) fn fs_stat(oid: UniOid, path: String) -> UniReturn<UniFsStat> {
    stat(oid_key(&oid), &path)
}

pub(super) fn fs_fsync(fd: u32) -> UniReturn<()> {
    let table = fd_table().read().expect("mock fs fd table lock poisoned");
    if !table.contains_key(&fd) {
        return error(ERRNO_BAD_F, format!("fs-fsync: unknown fd {fd}"));
    }
    // No-op: the emulation holds every byte in memory.
    Ok(())
}

pub(super) fn fs_readdir(oid: UniOid, path: String) -> UniReturn<Vec<UniFsDirent>> {
    let oid = oid_key(&oid);
    let files = files().read().expect("mock fs files lock poisoned");
    if files.contains_key(&(oid, path.clone())) {
        return error(ERRNO_NOT_DIR, format!("fs-readdir: '{path}' is a file"));
    }

    let prefix = if path.is_empty() {
        String::new()
    } else {
        format!("{}/", path.trim_end_matches('/'))
    };
    let mut entries: BTreeMap<String, UniFsDirent> = BTreeMap::new();
    for ((file_oid, entry), content) in files.iter() {
        if *file_oid != oid || !entry.starts_with(&prefix) {
            continue;
        }
        let rest = &entry[prefix.len()..];
        if rest.is_empty() {
            continue;
        }
        match rest.find('/') {
            Some(slash) => {
                let dir_name = &rest[..slash];
                entries
                    .entry(dir_name.to_string())
                    .or_insert_with(|| UniFsDirent {
                        name: dir_name.to_string(),
                        is_dir: true,
                        length: 0,
                    });
            }
            None => {
                entries.insert(
                    rest.to_string(),
                    UniFsDirent {
                        name: rest.to_string(),
                        is_dir: false,
                        length: content.len() as u64,
                    },
                );
            }
        }
    }
    Ok(entries.into_values().collect())
}
